rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,708 @@
+  "endLocation": {"gtfsId": "place-heath"},
+  "startTime": "04:47:00",
+  "endTime": "05:34:00"
+}
+```
+*Notes*:
+- The added/scheduled distinction here is internal to Glides, and may not align with what is in other data sources. For example, an Added Trip in Glides may still correspond to a scheduled GTFS trip, and a Scheduled Trip may not appear in GTFS (due to track closures, for example).
+
+### TripUpdated
+This indicates that a trip was updated in some fashion:  consist, operators, departure time. The trip is either a scheduled trip, or an added trip that has already appeared in the events stream. Fields which are not present are not considered to be updated.
+
+Fields in the object:
+- `type` (string): `"updated"|"added"`. Determines whether this is a TripUpdated object or a  [TripAdded](#TripAdded) object. If it's `"added"`, then it's a [TripAdded](#TripAdded) object, see above. Subsequent updates to previously-added trips have `type` `"updated"`.
+- `tripKey` ([TripKey](#TripKey)): which trip is being updated.
+- `comment` (string, optional): free text information about the trip. Could potentially be the empty string, if a comment was deleted.
+- `startLocation` ([Location](#Location) | "unset", optional): the new destination of the train.
+- `endLocation` ([Location](#Location) | "unset", optional): the new destination of the train.
+- `startTime` ([Time](#Time) | "unset", optional): if present, the new time that the train is expected to depart `startLocation` (or the existing `startLocation` of the trip).
+- `endTime` ([Time](#Time) | "unset", optional): if present, the new time that the train is expected to arrive at `endLocation` (or the existing `endLocation` of the trip).
+- `cars` (array of [Car](#Car), optional): array of length 1 or 2, containing the car numbers and operators for each car in the train assigned to the trip. If absent, there are no changes to any cars. If present, the length of the array is the length of the train. In a two car train, the front car is listed first.
+- `dropped` ([DroppedReason](#DroppedReason) | false, optional): If present the trip was dropped for the provided reason. If set to `false`, the trip is not dropped (and restored if previously dropped).
+- `scheduled` ([Scheduled](#Scheduled) | null): For trips in Glides' schedule, this represents the scheduled data for the trip. It will always be present and won't change between subsequent updates to the same trip. For added trips, this will always be `null`.
+
+Some values might be the special string `"unset"`. This means to treat the field as if it had never been modified. This is semantically slightly different than setting the value to the scheduled value. For example, setting `startTime` to `"unset"` implies we don't know when the trip will leave, and the scheduled time might be a good guess, but setting `startTime` to equal the scheduled start time implies that an inspector has confirmed the trip will leave at that time.
```

Comment:
To clarify a potential misunderstanding: Trips that Glides sees as scheduled won't get a `TripAdded` event the first time they're updated. They'll go straight into `TripUpdated` events.

(Even if we did use `TripAdded` the first time we sent an update to a scheduled trip, that wouldn't reduce any complexity for the client, because of the bit in [Short Term Storage](https://github.com/mbta/technology-docs/blob/ps-glides-trainsheet-events/rfcs/0018-glides-trainsheet-events.md#short-term-storage) "Clients MUST tolerate receiving updates for events that reference unseen past events (such as an update to an added trip that the client did not see get added)".)

---

"treat the field as if it had never been modified" means as if none of the past `TripAdded` or `TripUpdated` events set that field in their non-scheduled fields (scheduled data in `tripKey` and `scheduled` will be constant).

An example of how the semantics could matter for RTR:

Right now, RTR doesn't show predictions based on the schedule for countdown clocks at terminals, because they're not reliable enough. If an event was sent that said `{tripKey: {startTime: "10:00"}, startTime: "10:00"}` that means that an inspector has confirmed in Glides that the scheduled 10:00 trip will indeed start at 10:00. RTR should feel confident in showing a prediction on the countdown clock.

Then, the inspector says "wait I didn't mean to enter that" and deletes the data in Glides. An event is sent that says `{tripKey: {startTime: "10:00"}, startTime: "unset"}`. RTR should treat this like a trip that it knows is in the schedule for 10:00 but doesn't know anything else about. It should not feel confident enough to show it on the countdown clocks, and should bring the countdown clocks back to headway mode.